dist/tools/openocd/openocd.sh

```diff
@@ -353,11 +353,11 @@ do_debugserver() {
     # start OpenOCD as GDB server
     sh -c "${OPENOCD} \
             ${OPENOCD_ADAPTER_INIT} \
-            -f '${OPENOCD_CONFIG}' \
-            ${OPENOCD_EXTRA_INIT} \
             -c 'tcl_port ${TCL_PORT}' \
             -c 'telnet_port ${TELNET_PORT}' \
             -c 'gdb_port ${GDB_PORT}' \
+            -f '${OPENOCD_CONFIG}' \
+            ${OPENOCD_EXTRA_INIT} \
```

Comment:
Why is this inversion needed?

Reply:
ping?

Reply:
From what I can see its because the `OPENOCD_CONFIG` for this `BOARD` calls `init`

```
adapter speed 4000

adapter driver jlink
transport select jtag

set _CHIPNAME riscv
jtag newtap $_CHIPNAME cpu -irlen 5 -expected-id 0x20000913

set _TARGETNAME $_CHIPNAME.cpu
target create $_TARGETNAME.0 riscv -chain-position $_TARGETNAME
$_TARGETNAME.0 configure -work-area-phys 0x80000000 -work-area-size 0x4000 -work-area-backup 0

flash bank onboard_spi_flash fespi 0x20000000 0 0 0 $_TARGETNAME.0

init

jlink jtag 3

halt
flash protect 0 1 last off
echo "Ready for Remote Connections"
```

Reply:
Is the init actually needed? I replaced the config file with:

```
adapter speed 4000

adapter driver jlink
transport select jtag

set _CHIPNAME riscv
jtag newtap $_CHIPNAME cpu -irlen 5 -expected-id 0x20000913

set _TARGETNAME $_CHIPNAME.cpu
target create $_TARGETNAME.0 riscv -chain-position $_TARGETNAME
$_TARGETNAME.0 configure -work-area-phys 0x80000000 -work-area-size 0x4000 -work-area-backup 0

flash bank onboard_spi_flash fespi 0x20000000 0 0 0 $_TARGETNAME.0
```

Flashing debugging still works.

Reply:
I think we should not source `board` openocd configurations.

Reply:
> flash protect 0 1 last off

This line doesn't work for me:

```
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01383-gd46f28c2e-dirty (2020-08-20-10:53)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
flash
  flash bank bank_id driver_name base_address size_bytes chip_width_bytes
            bus_width_bytes target [driver_options ...]
  flash banks
  flash init
  flash list
gdb_flash_program ('enable'|'disable')
nand
program <filename> [address] [pre-verify] [verify] [reset] [exit]
Error: invalid subcommand "protect 0 1 last off"
/work/riot/RIOT/boards/hifive1b/dist/openocd.cfg:15: Error: 
in procedure 'script' 
at file "embedded:startup.tcl", line 26
at file "/work/riot/RIOT/boards/hifive1b/dist/openocd.cfg", line 15
Error: Target has not been initialized
```
It works if I remove it. I also reused the DEBUG_ADAPTER and OPENOCD_TRANSPORT variables.

Reply:
This is without the re-ordering you did in the openocd script.

Reply:
Yes, it's not needed when init is not called from the openocd.cfg file.

Reply:
I pushed an update that restore the `flash protect` command by using the `OPENOCD_PRE_FLASH_CMDS` variable. It still works:
```
make BOARD=hifive1b -C examples/default flash PROGRAMMER=openocd
make: Entering directory '/work/riot/RIOT/examples/default'
Building application "default" for "hifive1b" with MCU "fe310".

"make" -C /work/riot/RIOT/boards/hifive1b
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/fe310
"make" -C /work/riot/RIOT/cpu/fe310/nano
"make" -C /work/riot/RIOT/cpu/fe310/periph
"make" -C /work/riot/RIOT/cpu/fe310/vendor
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/drivers/saul
"make" -C /work/riot/RIOT/drivers/saul/init_devs
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/fmt
"make" -C /work/riot/RIOT/sys/isrpipe
"make" -C /work/riot/RIOT/sys/phydat
"make" -C /work/riot/RIOT/sys/ps
"make" -C /work/riot/RIOT/sys/saul_reg
"make" -C /work/riot/RIOT/sys/shell
"make" -C /work/riot/RIOT/sys/shell/commands
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  28474	    600	   2540	  31614	   7b7e	/work/riot/RIOT/examples/default/bin/hifive1b/default.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/examples/default/bin/hifive1b/default.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01383-gd46f28c2e-dirty (2020-08-20-10:53)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
jtag
Info : J-Link OB-K22-SiFive compiled Dec 12 2019 16:26:28
Info : Hardware version: 1.00
Info : VTarget = 3.300 V
Info : clock speed 4000 kHz
Info : JTAG tap: riscv.cpu tap/device found: 0x20000913 (mfg: 0x489 (SiFive, Inc.), part: 0x0000, ver: 0x2)
Info : datacount=1 progbufsize=16
Info : Disabling abstract command reads from CSRs.
Info : Examined RISC-V core; found 1 harts
Info :  hart 0: XLEN=32, misa=0x40101105
Info : starting gdb server for riscv.cpu.0 on 0
Info : Listening on port 42487 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* riscv.cpu.0        riscv      little riscv.cpu          running

Info : JTAG tap: riscv.cpu tap/device found: 0x20000913 (mfg: 0x489 (SiFive, Inc.), part: 0x0000, ver: 0x2)
Info : Found flash device 'issi is25lp032' (ID 0x0016609d)
cleared protection for sectors 1 through 63 on flash bank 0

Info : Disabling abstract command writes to CSRs.
auto erase enabled
wrote 65536 bytes from file /work/riot/RIOT/examples/default/bin/hifive1b/default.elf in 4.093731s (15.634 KiB/s)

verified 29076 bytes in 1.170141s (24.266 KiB/s)

Info : JTAG tap: riscv.cpu tap/device found: 0x20000913 (mfg: 0x489 (SiFive, Inc.), part: 0x0000, ver: 0x2)
shutdown command invoked
Done flashing
```

Reply:
Do you mind re-adding the `flash`, `debug`, `reset` output?

Reply:
flash: see above

<details><summary>debug</summary>

```
make BOARD=hifive1b -C examples/hello-world/ debug PROGRAMMER=openocd
make: Entering directory '/work/riot/RIOT/examples/hello-world'
/work/riot/RIOT/dist/tools/openocd/openocd.sh debug /work/riot/RIOT/examples/hello-world/bin/hifive1b/hello-world.elf
### Starting Debugging ###
Open On-Chip Debugger 0.10.0+dev-01383-gd46f28c2e-dirty (2020-08-20-10:53)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Reading symbols from /work/riot/RIOT/examples/hello-world/bin/hifive1b/hello-world.elf...
Remote debugging using :3333
0x2001088c in pm_set_lowest () at /work/riot/RIOT/cpu/fe310/periph/pm.c:26
26	    __asm__ volatile ("wfi");
(gdb) b main
Breakpoint 1 at 0x20010064: file /work/riot/RIOT/examples/hello-world/main.c, line 26.
(gdb) load
Loading section .init, size 0x64 lma 0x20010000
Loading section .text, size 0x2006 lma 0x20010064
Loading section .rodata, size 0x23c lma 0x2001206c
Loading section .sdata2._global_impure_ptr, size 0x4 lma 0x200122a8
Loading section .data, size 0x68 lma 0x200122ac
Start address 0x20010000, load size 8978
Transfer rate: 1 KB/sec, 1795 bytes/write.
(gdb) r
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /work/riot/RIOT/examples/hello-world/bin/hifive1b/hello-world.elf 
Note: automatically using hardware breakpoints for read-only addresses.

Breakpoint 1, main () at /work/riot/RIOT/examples/hello-world/main.c:26
26	    puts("Hello World!");
(gdb) l
21	
22	#include <stdio.h>
23	
24	int main(void)
25	{
26	    puts("Hello World!");
27	
28	    printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
29	    printf("This board features a(n) %s MCU.\n", RIOT_MCU);
30	
(gdb) quit
A debugging session is active.

	Inferior 1 [Remote target] will be detached.

Quit anyway? (y or n) y
Detaching from program: /work/riot/RIOT/examples/hello-world/bin/hifive1b/hello-world.elf, Remote target
[Inferior 1 (Remote target) detached]
```

</details>

<details><summary>reset</summary>

```
make BOARD=hifive1b -C examples/hello-world/ reset PROGRAMMER=openocd
make: Entering directory '/work/riot/RIOT/examples/hello-world'
/work/riot/RIOT/dist/tools/openocd/openocd.sh reset
### Resetting Target ###
Open On-Chip Debugger 0.10.0+dev-01383-gd46f28c2e-dirty (2020-08-20-10:53)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
jtag
Info : J-Link OB-K22-SiFive compiled Dec 12 2019 16:26:28
Info : Hardware version: 1.00
Info : VTarget = 3.300 V
Info : clock speed 4000 kHz
Info : JTAG tap: riscv.cpu tap/device found: 0x20000913 (mfg: 0x489 (SiFive, Inc.), part: 0x0000, ver: 0x2)
Info : datacount=1 progbufsize=16
Info : Disabling abstract command reads from CSRs.
Info : Examined RISC-V core; found 1 harts
Info :  hart 0: XLEN=32, misa=0x40101105
Info : Disabling abstract command writes to CSRs.
Info : [0] Found 8 triggers
Info : starting gdb server for riscv.cpu.0 on 0
Info : Listening on port 44123 for gdb connections
Info : JTAG tap: riscv.cpu tap/device found: 0x20000913 (mfg: 0x489 (SiFive, Inc.), part: 0x0000, ver: 0x2)
shutdown command invoked
make: Leaving directory '/work/riot/RIOT/examples/hello-world'
```

</details>